File: src/day-03/rucksack-reorganization.clj
(ns aoc22.day-03.rucksack-reorganization
  (:require [clojure.string :as s]
            [clojure.set :as set]))

(def ex-inp (str "vJrwpWtwJgWrhcsFMMfFFhFp\n"
                 "jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL\n"
                 "PmmdzqPrVvPwwTWBwg\n"
                 "wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn\n"
                 "ttgJtRGJQctTZtZT\n"
                 "CrZsJsPPZsGzwwsLwLmpwMDw"))

(def inp (slurp "src/day-03/inp.txt"))

(def points (into {} (mapcat (fn [i] [[(char (+ i (int \a))) (+ i 1)]
                                      [(char (+ i (int \A))) (+ i 27)]])
                             (range 0 26))))

(defn ex1 [text]
  (->> (s/split-lines text)
       (reduce (fn [acc line]
                 (->> (-> (count line) (quot 2) (split-at (seq line)))
                      (map set)
                      (reduce set/intersection)
                      (first)
                      (points)
                      (+ acc))) 0)))

(defn ex2 [text]
  (->> (s/split-lines text)
       (partition 3)
       (reduce (fn [acc elves-group]
                 (let [[elf1 elf2 elf3] (map set elves-group)
                       inter (set/intersection elf1 elf2 elf3)]
                   (+ acc (points (first inter))))) 0)))


(ex1 ex-inp) ;; 157
(ex1 inp) ;; 7980
(ex2 ex-inp) ;; 70
(ex2 inp) ;; 2881

File: elixir/lib/day-03/rucksack_reorganization.ex
defmodule RucksackReorganizations do
  @priorities 0..25
              |> Enum.flat_map(fn i ->
                char_to_string = &List.to_string([&1])
                [{char_to_string.(?a + i), i + 1}, {char_to_string.(?A + i), i + 27}]
              end)
              |> Enum.into(%{})

  def ex1(text) do
    text
    |> String.split("\n")
    |> Enum.reduce(0, fn line, acc ->
      half = line |> String.length() |> div(2)
      {item1, item2} = String.split_at(line, half)
      priority = intersection_priority([item1, item2])

      @priorities[priority] + acc
    end)
  end

  def ex2(text) do
    text
    |> String.split("\n")
    |> Enum.chunk_every(3)
    |> Enum.reduce(0, &(@priorities[intersection_priority(&1)] + &2))
  end

  defp intersection_priority(list) do
    list
    |> Enum.map(&MapSet.new(String.codepoints(&1)))
    |> Enum.reduce(&MapSet.intersection/2)
    |> MapSet.to_list()
    |> List.first()
  end
end

ex_inp = File.read!("./lib/day-03/ex-inp.txt")
inp = File.read!("./lib/day-03/inp.txt")

ex_inp |> RucksackReorganizations.ex1() |> IO.puts()
inp |> RucksackReorganizations.ex1() |> IO.puts()
ex_inp |> RucksackReorganizations.ex2() |> IO.puts()
inp |> RucksackReorganizations.ex2() |> IO.puts()
